conn.go

```diff
@@ -347,6 +380,93 @@ func (conn *Conn) inWorker() {
 	}
 }
 
+// sigWorker gets signals from signal queue and handles them
+// appropriately.
+func (conn *Conn) sigWorker() {
+	for {
+		signal := conn.nextSignal()
+		if signal != nil {
+			closing := conn.handleSignal(signal)
+			if closing {
+				return
+			}
+		} else {
+			select {
+			case <-conn.signalQueueWakeUp:
+			case <-conn.shutdownSigWorker:
+				return
+			}
+		}
+	}
+}
+
+// nextSignal gets next signal from queue or nil if there was no
+// signal in queue.
+func (conn *Conn) nextSignal() *Signal {
+	conn.signalQueueLck.Lock()
+	defer conn.signalQueueLck.Unlock()
+	for len(conn.signalQueue) > 0 {
+		signal := conn.signalQueue[0]
+		conn.signalQueue = conn.signalQueue[1:]
+		if signal != nil {
+			return signal
+		}
+	}
+
+	conn.signalQueueConsumerAsleep = true
+	return nil
+}
+
+// processPendingSignalChannels either adds new pending channels to
+// signal channels list or removes them if they already existed there.
+//
+// It is assumed that conn.signalsLck is locked.
+func (conn *Conn) processPendingSignalChannels() {
+	signalsMap := make(map[chan<- *Signal]bool, len(conn.signals))
+	for _, ch := range conn.signals {
+		signalsMap[ch] = false
+	}
+	conn.pendingSignalChannelsLck.Lock()
+	for _, ch := range conn.pendingSignalChannels {
+		if _, ok := signalsMap[ch]; ok {
+			delete(signalsMap, ch)
+		} else {
+			signalsMap[ch] = false
+		}
+	}
+	conn.pendingSignalChannels = nil
+	conn.pendingSignalChannelsLck.Unlock()
+
+	conn.signals = nil
+	for ch, _ := range signalsMap {
+		conn.signals = append(conn.signals, ch)
```

Comment:
Now I don't follow. Copy a channel?

But in general assumption is that if you pass a channel to conn.Signal then conn takes it over (it closes it in conn.Close). You later just use it for listening to channels.